### DISCUSSION TITLE
Can I programmatically calculate a hash of "public" parts of the dll?

### DISCUSSION THREAD

**Discussion by safesparrow** [icsharpcode/ILSpy#2937] [Help] [answered]
Hi,

I would like to calculate an assembly's footprint - a hash that changes only if "significant" parts of the assembly change, which is mostly when its [Reference assemblies](https://learn.microsoft.com/en-us/dotnet/standard/assembly/reference-assemblies) would be different.
This is to help with incremental build (through custom MSBuild targets).

Some assemblies produce Reference assemblies already, so for those I can just check the hash of the ref dll.
Reference assemblies have dummy method implementation, but most of the assembly structure stays the same.
However, some other assemblies don't produce ref assemblies (eg. many NuGet packages, most F# projects (as it's a new feature in F#)).
Therefore I'd like to construct such a hash based on the original dll, ideally in memory. It would need to be fairly fast, as its aim is to be much faster than the actual compilation of a downstream project.

Can I use `ICSharpCode.Decompiler` to achieve this (I know it has `Decompiler` in the name 🙂)?
So far I looked at https://github.com/icsharpcode/ILSpy/blob/master/decompiler-nuget-demos.ipynb , but it's not obvious to me:
1. whether it allows modifying the entities it returns
2. whether there is any way to reconstruct the dll from those entities (so that I can then calculate a hash of that dll)
3. whether there is any way to generate a hash from those entities directly

Is there an alternative programmatic tool I could use?

Thanks

**Comment by dgrunwald** [comment, marked as answer]
ILSpy uses System.Reflection.Metadata as the low-level assembly loader. This provides a read-only view into the .dll, no way to modify it.
At a higher level, ILSpy has its own type system (ICSharpCode.Decompiler.TypeSystem). This again is read-only.

If you want to modify a .dll, remove all method bodies and maybe the private methods, save it to a MemoryStream, and then hash it; then I would recommend using the Mono.Cecil library.

If you want the fastest assembly-API-hashing tool possible, and have some flexibility in what exactly you are hashing, then I would recommend using System.Reflection.Metadata with a custom "signature decoder" that directly adds to a hasher. Then, for public classes and methods, write the relevant fields from the SRM structs to your hasher; while "decoding" signature blobs (you can't hash signature blobs directly as they contain metadata tokens which contain row numbers that might get re-numbered whenever another private method/class is added). But this will be **significantly** more work than the Mono.Cecil solution.
